src/main/java/frc/lib/posestimator/PoseEstimator.java

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (C) 2025 Windham Windup
+ *
+ * This program is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program. If
+ * not, see <https://www.gnu.org/licenses/>.
+ */
+
+package frc.lib.posestimator;
+
+import static edu.wpi.first.units.Units.Seconds;
+
+import java.util.Optional;
+import edu.wpi.first.apriltag.AprilTagFieldLayout;
+import edu.wpi.first.math.Matrix;
+import edu.wpi.first.math.Nat;
+import edu.wpi.first.math.VecBuilder;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Transform2d;
+import edu.wpi.first.math.geometry.Twist2d;
+import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
+import edu.wpi.first.math.numbers.N3;
+import edu.wpi.first.units.measure.Time;
+
+import frc.lib.io.vision.VisionIO.VisionObservation;
+import frc.lib.posestimator.PoseEstimator.VisionProcessor.PoseRecord;
+import frc.lib.posestimator.SwerveOdometry.OdometryObservation;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.experimental.Accessors;
+
+/**
+ * Estimates the robot's field-relative pose using a combination of swerve-drive odometry and
+ * AprilTag-based vision observations.
+ *
+ * <p>
+ * This class fuses kinematic odometry data with visual pose estimates to produce a more accurate
+ * and robust {@link Pose2d} estimate of the robot's position and orientation on the field. It
+ * supports single- and multi-tag vision measurements, distance-based confidence scaling, and
+ * single-tag based triangulation for zero-ambiguity pose estimation with close distances.
+ *
+ * <p>
+ * Pose updates occur via two main data sources:
+ * <ul>
+ * <li>{@link #addOdometryObservation(OdometryObservation)} — incremental odometry updates</li>
+ * <li>{@link #addVisionObservation(VisionObservation)} — vision-based pose corrections</li>
+ * </ul>
+ *
+ * <p>
+ * Recent odometry samples are also buffered to support timestamp interpolation.
+ */
+@Accessors(fluent = true)
+public class PoseEstimator {
+    public static interface VisionProcessor {
```

Comment:
nit: This probably deserves to be in its own file, in the same package as the implementations